### ISSUE TITLE
Scheduler crashes.

### ISSUE THREAD

**Issue by Ramyak** [kubernetes/kubernetes#110726]
### What happened?

Scheduler went into crash loop.

We upgraded from 1.18 => 1.19 and scheduler started crashing few 100 times and then auto-recovered.

Probably related to this code change - https://github.com/kubernetes/kubernetes/commit/65d9f8175844

Seems like we no longer use pointers in HEAD
https://github.com/kubernetes/kubernetes/pull/107623

1.19 - 1.23 has this bug. Fixed 1.24 onwards.

File corresponding to this crash.
https://github.com/kubernetes/kubernetes/blob/v1.19.16/pkg/scheduler/framework/plugins/podtopologyspread/filtering.go#L139

```
2022-06-16T23:52:07.604670059+00:00 stderr F I0616 23:52:07.604621       1 scheduler.go:456] Attempting to schedule pod: kube-detection/redcanary-sensor-productio
n-n4csf
2022-06-16T23:52:07.606291246+00:00 stderr F E0616 23:52:07.606257       1 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (r
untime error: invalid memory address or nil pointer dereference)
2022-06-16T23:52:07.606291246+00:00 stderr F goroutine 1985434 [running]:
2022-06-16T23:52:07.606291246+00:00 stderr F k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime.logPanic(0x1a9df00, 0x2cd6d60)
2022-06-16T23:52:07.606291246+00:00 stderr F    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/
runtime.go:74 +0xa6
2022-06-16T23:52:07.606291246+00:00 stderr F k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
2022-06-16T23:52:07.606291246+00:00 stderr F    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:48 +0x89
2022-06-16T23:52:07.606291246+00:00 stderr F panic(0x1a9df00, 0x2cd6d60)
2022-06-16T23:52:07.606291246+00:00 stderr F    /usr/local/go/src/runtime/panic.go:969 +0x1b9
2022-06-16T23:52:07.606291246+00:00 stderr F k8s.io/kubernetes/pkg/scheduler/framework/plugins/podtopologyspread.(*preFilterState).updateWithPod(0xc049b17530, 0xc03f92d800, 0xc03be79800, 0xc01abf5500, 0x1)
2022-06-16T23:52:07.606291246+00:00 stderr F    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/scheduler/framework/plugins/podtopologyspread/filtering.go:139 +0x285
2022-06-16T23:52:07.606291246+00:00 stderr F k8s.io/kubernetes/pkg/scheduler/framework/plugins/podtopologyspread.(*PodTopologySpread).AddPod(0xc000b78900, 0x1f8d900, 0xc03b236240, 0xc049b17500, 0xc03be79800, 0xc03f92d800, 0xc03546fef0, 0x0)
2022-06-16T23:52:07.606291246+00:00 stderr F    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/scheduler/framework/plugins/podtopologyspread/filtering.go:167 +0x7f
2022-06-16T23:52:07.606291246+00:00 stderr F k8s.io/kubernetes/pkg/scheduler/framework/runtime.(*frameworkImpl).runPreFilterExtensionAddPod(0xc0004aa820, 0x1f8d900, 0xc03b236240, 0x7fdb58c5bbf8, 0xc000b78900, 0xc049b17500, 0xc03be79800, 0xc03f92d800, 0xc03546fef0, 0x0)
2022-06-16T23:52:07.606291246+00:00 stderr F    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/scheduler/framework/runtime/framework.go:447 +0xb4
2022-06-16T23:52:07.606291246+00:00 stderr F k8s.io/kubernetes/pkg/scheduler/framework/runtime.(*frameworkImpl).RunPreFilterExtensionAddPod(0xc0004aa820, 0x1f8d900, 0xc03b236240, 0xc049b17500, 0xc03be79800, 0xc03f92d800, 0xc03546fef0, 0x0)
2022-06-16T23:52:07.606291246+00:00 stderr F    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/scheduler/framework/runtime/framework.go
:433 +0x137
2022-06-16T23:52:07.606291246+00:00 stderr F k8s.io/kubernetes/pkg/scheduler/core.addNominatedPods(0x1f8d900, 0xc03b236240, 0x1fa0e00, 0xc000136b80, 0xc03be79800,
 0xc0287d9560, 0xc04a66d290, 0xc026310300, 0xe90000c0287d9560, 0xc03aab3ef0, ...)
2022-06-16T23:52:07.606291246+00:00 stderr F    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/scheduler/core/generic_scheduler.go:415
+0x1f7
2022-06-16T23:52:07.606291246+00:00 stderr F k8s.io/kubernetes/pkg/scheduler/core.PodPassesFiltersOnNode(0x1f8d900, 0xc03b236240, 0x1fa0e00, 0xc000136b80, 0xc0287
d9560, 0xc03be79800, 0xc04a66d290, 0xe3f8fcdabf00, 0xc037abddc0, 0x0, ...)
d9560, 0xc03be79800, 0xc04a66d290, 0xe3f8fcdabf00, 0xc037abddc0, 0x0, ...)
2022-06-16T23:52:07.606291246+00:00 stderr F    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/scheduler/core/generic_scheduler.go:469
+0x23c
2022-06-16T23:52:07.606291246+00:00 stderr F k8s.io/kubernetes/pkg/scheduler/core.(*genericScheduler).findNodesThatPassFilters.func1(0x18a)
2022-06-16T23:52:07.606291246+00:00 stderr F    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/scheduler/core/generic_scheduler.go:322
+0x159
2022-06-16T23:52:07.606291246+00:00 stderr F k8s.io/kubernetes/vendor/k8s.io/client-go/util/workqueue.ParallelizeUntil.func1(0xc0263103e0, 0xc039004240, 0xc026310
3d0, 0x1e3, 0xc000d14100, 0xc030ade000)
2022-06-16T23:52:07.606291246+00:00 stderr F    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/util/workqueue/paral
lelizer.go:90 +0x90
2022-06-16T23:52:07.606291246+00:00 stderr F created by k8s.io/kubernetes/vendor/k8s.io/client-go/util/workqueue.ParallelizeUntil
2022-06-16T23:52:07.606291246+00:00 stderr F    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/util/workqueue/paral
lelizer.go:76 +0x253
2022-06-16T23:52:07.609485909+00:00 stderr F panic: runtime error: invalid memory address or nil pointer dereference [recovered]
2022-06-16T23:52:07.609485909+00:00 stderr P    panic:
2022-06-16T23:52:07.609497381+00:00 stderr F runtime error: invalid memory address or nil pointer dereference
2022-06-16T23:52:07.609497381+00:00 stderr F [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x182ab25]
2022-06-16T23:52:07.609497381+00:00 stderr F
2022-06-16T23:52:07.609497381+00:00 stderr P goroutine 1985434 [running
2022-06-16T23:52:07.609505188+00:00 stderr F ]:
2022-06-16T23:52:07.609505188+00:00 stderr F k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
2022-06-16T23:52:07.609512215+00:00 stderr F    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/
runtime.go:55 +0x10c
2022-06-16T23:52:07.609512215+00:00 stderr P panic
2022-06-16T23:52:07.609523806+00:00 stderr F (0x1a9df00, 0x2cd6d60)
2022-06-16T23:52:07.609523806+00:00 stderr F    /usr/local/go/src/runtime/panic.go:969 +0x1b9
2022-06-16T23:52:07.609523806+00:00 stderr P k8s.io/kubernetes/pkg/scheduler/framework/plugins/podtopologyspread.(*preFilterState).updateWithPod(0xc049b17530
2022-06-16T23:52:07.609531398+00:00 stderr F , 0xc03f92d800, 0xc03be79800, 0xc01abf5500, 0x1)
2022-06-16T23:52:07.609531398+00:00 stderr P    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/scheduler/framework/plugins/podtopologys
pread/filtering.go
2022-06-16T23:52:07.609538356+00:00 stderr F :139 +0x285
2022-06-16T23:52:07.609538356+00:00 stderr P k8s.io/kubernetes/pkg/scheduler/framework/plugins/podtopologyspread.(*PodTopologySpread).AddPod(0xc000b78900,
2022-06-16T23:52:07.609544978+00:00 stderr P 0x1f8d900, 0xc03b236240, 0xc049b17500, 0xc03be79800, 0xc03f92d800,
2022-06-16T23:52:07.609551604+00:00 stderr F 0xc03546fef0, 0x0)
2022-06-16T23:52:07.609551604+00:00 stderr P    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/scheduler/framework/plugins/podtopologys
pread/filtering.go:167 +0x7f
2022-06-16T23:52:07.609558101+00:00 stderr F
2022-06-16T23:52:07.609558101+00:00 stderr P k8s.io/kubernetes/pkg/scheduler/framework/runtime.(*frameworkImpl).runPreFilterExtensionAddPod(0xc0004aa820, 0x1f8d90
0, 0xc03b236240,
2022-06-16T23:52:07.609564508+00:00 stderr P 0x7fdb58c5bbf8, 0xc000b78900, 0xc049b17500, 0xc03be79800, 0xc03f92d800
2022-06-16T23:52:07.609570774+00:00 stderr F , 0xc03546fef0, 0x0)
2022-06-16T23:52:07.609570774+00:00 stderr P    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/scheduler/framework/runtime/framework.go
:447
2022-06-16T23:52:07.609577228+00:00 stderr F  +0xb4
2022-06-16T23:52:07.609577228+00:00 stderr P k8s.io/kubernetes/pkg/scheduler/framework/runtime.(*frameworkImpl).RunPreFilterExtensionAddPod(0xc0004aa820, 0x1f8d90
0,
2022-06-16T23:52:07.609583641+00:00 stderr P 0xc03b236240, 0xc049b17500, 0xc03be79800, 0xc03f92d800, 0xc03546fef0,
2022-06-16T23:52:07.609589911+00:00 stderr F 0x0)
2022-06-16T23:52:07.609589911+00:00 stderr F    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/scheduler/framework/runtime/framework.go
:433 +0x137
2022-06-16T23:52:07.609596303+00:00 stderr P k8s.io/kubernetes/pkg/scheduler/core.addNominatedPods(0x1f8d900, 0xc03b236240, 0x1fa0e00, 0xc000136b80
2022-06-16T23:52:07.609603151+00:00 stderr P , 0xc03be79800, 0xc0287d9560, 0xc04a66d290, 0xc026310300, 0xe90000c0287d9560
2022-06-16T23:52:07.609609483+00:00 stderr F , 0xc03aab3ef0, ...)
2022-06-16T23:52:07.609609483+00:00 stderr P    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/scheduler/core/generic_scheduler.go:415
+
2022-06-16T23:52:07.609615945+00:00 stderr F 0x1f7
2022-06-16T23:52:07.609615945+00:00 stderr P k8s.io/kubernetes/pkg/scheduler/core.PodPassesFiltersOnNode(0x1f8d900, 0xc03b236240,
2022-06-16T23:52:07.609622433+00:00 stderr P 0x1fa0e00, 0xc000136b80, 0xc0287d9560, 0xc03be79800, 0xc04a66d290,
2022-06-16T23:52:07.609628702+00:00 stderr F 0xe3f8fcdabf00, 0xc037abddc0, 0x0, ...)
2022-06-16T23:52:07.609628702+00:00 stderr P    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/scheduler/core/generic_scheduler.go
2022-06-16T23:52:07.609635337+00:00 stderr F :469 +0x23c
2022-06-16T23:52:07.609635337+00:00 stderr F k8s.io/kubernetes/pkg/scheduler/core.(*genericScheduler).findNodesThatPassFilters.func1(0x18a)
2022-06-16T23:52:07.609641735+00:00 stderr F    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/scheduler/core/generic_scheduler.go:322
+0x159
2022-06-16T23:52:07.609641735+00:00 stderr P k8s.io/kubernetes/vendor/k8s.io/client-go/util/workqueue.ParallelizeUntil.func1(
2022-06-16T23:52:07.609648238+00:00 stderr P 0xc0263103e0, 0xc039004240, 0xc0263103d0, 0x1e3, 0xc000d14100,
2022-06-16T23:52:07.609654586+00:00 stderr F 0xc030ade000)
2022-06-16T23:52:07.609654586+00:00 stderr F    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/util/workqueue/paral
lelizer.go:90 +0x90
2022-06-16T23:52:07.609663576+00:00 stderr F created by k8s.io/kubernetes/vendor/k8s.io/client-go/util/workqueue.ParallelizeUntil
2022-06-16T23:52:07.609663576+00:00 stderr F    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/util/workqueue/paral
lelizer.go:76 +0x253
```

### What did you expect to happen?

Scheduler not to crash.

### How can we reproduce it (as minimally and precisely as possible)?

* Enable pod topology spread on multiple deployments.
* Do a large deployment - so that `addNominatedPods` gets called.
   ```
  Schedule
  findNodesThatFitPod
  findNodesThatPassFilters
  PodPassesFiltersOnNode
  addNominatedPods
  RunPreFilterExtensionAddPod
  runPreFilterExtensionAddPod
  AddPod
  ```
* Scheduler crashes every few seconds until pending pods drop down to 0.

### Kubernetes version

<details>

```console
$ kubectl --context=m-c-ea1-us version
Client Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.1", GitCommit:"632ed300f2c34f6d6d15ca4cef3d3c7073412212", GitTreeState:"clean", BuildDate:"2021-08-19T15:38:26Z", GoVersion:"go1.16.6", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"19+", GitVersion:"v1.19.16-airbnb0", GitCommit:"d6d1d6aa0b20c3c5e02bd7b41ee7d3e73cf7cad9", GitTreeState:"clean", BuildDate:"2022-01-11T07:36:32Z", GoVersion:"go1.15.15", Compiler:"gc", Platform:"linux/amd64"}
```

</details>

### Cloud provider

<details>
Self hosted
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
NAME="Amazon Linux"
VERSION="2"
ID="amzn"
ID_LIKE="centos rhel fedora"
VERSION_ID="2"
PRETTY_NAME="Amazon Linux 2"
ANSI_COLOR="0;33"
CPE_NAME="cpe:2.3:o:amazon:amazon_linux:2"
HOME_URL="https://amazonlinux.com/"

$ uname -a
Linux i-0b0423ff11f766555 5.10.102-99.473.amzn2.x86_64 #1 SMP Wed Mar 2 19:14:12 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by kerthcet** [comment]
It seems like we can't find the kv pair in `s.TpPairToMatchNum`, do we have new nodes added to the cluster? https://github.com/kubernetes/kubernetes/blob/e37e4ab4cc8dcda84f1344dda47a97bb1927d074/pkg/scheduler/framework/plugins/podtopologyspread/filtering.go#L138

**Comment by Ramyak** [comment]
Yes. The nodes are in an autoscaling group. So nodes get added/removed all the time.

We are going to try to apply this patch to see if this avoid the crash from 1.19 - 1.23

```
+               klog.V(1).Infof("hii s.TpPairToMatchNum '%v' pair: '%v'", s.TpPairToMatchNum, pair)
+               if s.TpPairToMatchNum == nil {
+                       klog.V(1).Infof("hii avoided crash pair: '%v'", pair)
+                       continue
+               }
```

**Comment by Ramyak** [comment]
Yeah make sense.. will try that. Thanks.

**Comment by Ramyak** [comment]
Yes it fixed the issue. `s.TpPairToMatchNum[pair]`  being `nil` was the problem. 
```
2022-06-23T23:50:42.356465338+00:00 stderr F I0623 23:50:42.356454       1 filtering.go:144] hii avoided crash pair: '{topology.kubernetes.io/zone us-east-1a}'
2022-06-23T23:50:43.355794208+00:00 stderr F I0623 23:50:43.355791       1 filtering.go:144] hii avoided crash pair: '{topology.kubernetes.io/zone us-east-1a}'
2022-06-23T23:50:54.500357803+00:00 stderr F I0623 23:50:54.500295       1 filtering.go:144] hii avoided crash pair: '{topology.kubernetes.io/zone us-east-1a}'
```


We are going to roll it out more widely tomorrow.

```
+               klog.V(1).Infof("hii s.TpPairToMatchNum '%v' pair: '%v'", s.TpPairToMatchNum, pair)
+               if s.TpPairToMatchNum[pair] == nil {
+                       klog.V(1).Infof("hii avoided crash pair: '%v'", pair)
+                       s.TpPairToMatchNum[pair] = new(int32)
+               }
```

**Comment by kerthcet** [comment]
It's still strange since we'll snapshot the nodeList at the beginning of every scheduling cycle, so in theory, we won't miss the `{topology.kubernetes.io/zone us-east-1a}` pair, can you reproduce this with 1.22 or 1.23, 1.19 is out of community support now.

**Comment by kerthcet** [comment]
This could happen as below:
1. We have two pods: P1, P2. P1 is the same as P2 except the pod name
2. We have Node1 fit(nodeAffinity and taint tolerations) for both P1 and P2
3. Try to schedule P1, Node1 is resource exhausted, but P1 preempted successfully.
4. We update Node1 with labels not fit for P2 any more, e.g. P2 is configured with `nodeAffinityNotIn` labelA, then I marked nodeA with this labelA.
5. Try to schedule P2, in PodTopologySpread plugin, since NodeA is no longer fit any more, so we skip the calculation of `TpPairToMatchNum `
6. Then we ran to `RunFilterPluginsWithNominatedPods` -> `addNominatedPods` -> `RunPreFilterExtensionAddPod` -> `AddPod` -> `updateWithPod`, we'd like to add nominated pod `P1`, but we don't verify whether `P1` meet the `nodeAffinity` or not, then we'll call `s.TpPairToMatchNum[pair]`, it will return nil. That's the problem
7. In releasev1.24, we changed the field type of `TpPairToMatchNum` to `int`, this problem is buried.

ToDo:

- [ ] Fix the bug in master, I created a PR already, but missed some tests.
- [ ] Hotfix the bug in lower releases

Whatever, when we call `AddPod` or `RemovePod` in PodTopologySpread, we should verify whether the node is satisfied.  
cc sig-scheduling team.

**Comment by Ramyak** [comment]
> can you reproduce this with 1.22 or 1.23, 1.19 is out of community support now.

No. I only have 1.19 running in prod..

**Comment by kerthcet** [comment]
> No. I only have 1.19 running in prod..

It doesn't matter, it's actually a bug I think.

**Comment by jiancheung** [comment]
Oh to clarify, we (with @Ramyak ) saw this exact crash in both a 1.19 and a 1.23 cluster. We only tested and verify the fix on 1.19 clusters though.

**Comment by ahg-g** [comment]
> It's still strange since we'll snapshot the nodeList at the beginning of every scheduling cycle

The call to updateWithPod is being made in a different scheduling cycle (it is for the incoming pod) than the one that nominated the pod. Your analysis in https://github.com/kubernetes/kubernetes/issues/110726#issuecomment-1165513276 makes sense and is consistent with the fact that we snapshot the nodes for every scheduling cycle.

**Comment by kerthcet** [comment]
Thanks @ahg-g for feedbacks, and I think we should hotfix this bug in lower releases, WDYT?

**Comment by kerthcet** [comment]
Just created a hotfix PR https://github.com/kubernetes/kubernetes/pull/110853, when this approved, I'll cherry-pick to other releases.

**Comment by ingvagabund** [comment]
Hey @kerthcet, thank You for fixing the crash in [master](https://github.com/kubernetes/kubernetes/pull/110768) and [release-1.23](https://github.com/kubernetes/kubernetes/pull/110853). It looks like the fix actually missed [release-1.24 branch](https://github.com/kubernetes/kubernetes/blob/release-1.24/pkg/scheduler/framework/plugins/podtopologyspread/filtering.go#L150-L170). Would you please backport the fix into release-1.24 as well?

**Comment by alculquicondor** [comment]
/close
this is fixed
